File: subly_android/app/src/main/java/net/cynreub/subly/domain/repository/CategoryRepository.kt
package net.cynreub.subly.domain.repository

import kotlinx.coroutines.flow.Flow
import net.cynreub.subly.domain.model.Category
import net.cynreub.subly.ui.categories.CategoryWithCount
import java.util.UUID

interface CategoryRepository {
    fun getAllCategories(): Flow<List<Category>>
    fun getAllCategoriesWithCount(): Flow<List<CategoryWithCount>>
    fun getCategoryById(id: UUID): Flow<Category?>
    suspend fun getUsageCount(id: UUID): Int
    suspend fun insertCategory(category: Category)
    suspend fun updateCategory(category: Category)
    /** @throws IllegalStateException if the category is used by one or more subscriptions */
    suspend fun deleteCategory(category: Category)
}

File: subly_ios/Subly/Subly/Domain/Repositories/CategoryRepository.swift
import Foundation
import Combine

protocol CategoryRepository {
    func getAllCategories() -> AnyPublisher<[Category], Never>
    func getAllCategoriesWithCount() -> AnyPublisher<[CategoryWithCount], Never>
    func getCategoryById(_ id: UUID) -> AnyPublisher<Category?, Never>
    func getUsageCount(for id: UUID) async -> Int
    func insertCategory(_ category: Category) async throws
    func updateCategory(_ category: Category) async throws
    /// Throws if the category is used by one or more subscriptions
    func deleteCategory(_ category: Category) async throws
}
